proxy/src/telemetry/metrics/mod.rs

```diff
@@ -211,17 +268,80 @@ impl Metrics {
             .entry(labels.clone())
             .or_insert_with(Default::default)
     }
-}
 
+    fn tcp_accept_open_total(&mut self,
+                             labels: &Arc<TransportLabels>)
+                             -> &mut Counter {
+        self.tcp_accept_open_total.values
+            .entry(labels.clone())
+            .or_insert_with(Default::default)
+    }
+
+    fn tcp_accept_close_total(&mut self,
+                              labels: &Arc<TransportLabels>)
+                              -> &mut Counter {
+        self.tcp_accept_close_total.values
+            .entry(labels.clone())
+            .or_insert_with(Default::default)
+    }
+
+    fn tcp_connect_open_total(&mut self,
+                             labels: &Arc<TransportLabels>)
+                             -> &mut Counter {
+        self.tcp_connect_open_total.values
+            .entry(labels.clone())
+            .or_insert_with(Default::default)
+    }
+
+    fn tcp_connect_close_total(&mut self,
+                              labels: &Arc<TransportLabels>)
+                              -> &mut Counter {
+        self.tcp_connect_close_total.values
+            .entry(labels.clone())
+            .or_insert_with(Default::default)
+    }
+
+    fn tcp_connection_duration(&mut self,
+                                labels: &Arc<TransportLabels>)
+                                -> &mut Histogram {
+        self.tcp_connection_duration.values
+            .entry(labels.clone())
+            .or_insert_with(Default::default)
+    }
+
+    fn sent_bytes(&mut self,
+                  labels: &Arc<TransportLabels>)
+                  -> &mut Counter {
+        self.sent_bytes.values
+            .entry(labels.clone())
+            .or_insert_with(Default::default)
+    }
+
+    fn received_bytes(&mut self,
+                      labels: &Arc<TransportLabels>)
+                      -> &mut Counter {
+        self.received_bytes.values
+            .entry(labels.clone())
+            .or_insert_with(Default::default)
+    }
+}
 
 impl fmt::Display for Metrics {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}\n{}\n{}\n{}\n{}\nprocess_start_time_seconds {}\n",
+        write!(f,
```

Comment:
I can't decide if this change is for the better, just seeing `{}\n{}\n{}\n`... made me first react "I should count there are enough!", until I realized the compiler will do that for me XD

What do you think?

```rust
writeln!(f, "{}", self.request_total)?;
writeln!(f, "{}", self.request_duration)?;
writeln!(f, "{}", self.response_total)?;
writeln!(f, "{}", self.response_duration)?;

// etc ...
```

Reply:
I could go either way, though it does appear that there might be a functional difference in error behaviour between the current code & what you posted --- do you think it's better to try and write out the metrics piece by piece?

Reply:
I don't have a strong opinion either way, and so kind of didn't want to leave the comment, but figured I'd offer the alternative in case you hadn't considered it.

What functional difference do you see? It should actually be identical. The current code will still look for an error after formatting each `{}`...

Reply:
Ah, you're right, never mind --- it should be equivalent.